### ISSUE TITLE
`Lists Should Be Equal` keywords in Collections should have the option to ignore order

### ISSUE THREAD

**Issue by bkhouri** [robotframework/robotframework#2703]
when comparing lists, we sometimes want to compare only the content irregardless of the order.

Currently, the `Lists should be equal` ensures each index in the list is the same.

So the following test case fails

```
*** Settings ***
Library     Collections

*** Test Cases ***
Compare lists
    ${l1}=  create list     a   b    c
    ${l2}=  create list     c   a    b
    lists should be equal   ${l1}   ${l2}
```

If I wanted to get this test to pass, I would need to use a combination of keywords found in `Collections` and `BuiltIn`.

```
*** Settings ***
Library     Collections

*** Test Cases ***
Compare lists
*** Settings ***
    ${l1}=  create list     a   b    c
    ${l2}=  create list     c   a    b
    list should contain sub list  ${l1}     ${l2}
    ${l1_length}=       get length  ${l1}
    length should be    ${l2}       ${l1_length}

```

I would be nice to be able to do something as follows:

```
*** Settings ***
Library     Collections

*** Test Cases ***
Compare lists
    ${l1}=  create list     a   b    c
    ${l2}=  create list     c   a    b
    lists should be equal   ${l1}   ${l2}      ignore_order=True
```

This ticket has been opened to improve the `Lists should be equal` keywords to include a way to ignore the orders of the list.

**Comment by pekkaklarck** [comment]
Sounds reasonable. What about duplicates?

**Comment by bkhouri** [comment]
I haven't given some thoughts about duplicates, but I guess we can have an "ignore_duplicates" argument as well.

**Comment by bbpatel2001** [comment]
This `Lists should be equal` keyword first compares  length of two list and if it is not same it fails. 
if we ignore_order=True as argument, we can check first list's element in 2nd list  and if it is available at any position, test should pass.  

**Pseudo code** 
If  ignore_order=True
    remove duplicates from lists
    Fail if  not same length
    if  element is not available in 2nd list, assert fail.
else  ##  order should be checked 
    compare item1 and item2

```
if ignore_order == True: 
    assert_true(item1 in list2, msg='Index %d%s' % (index,  " Not in second list"))
else:
    assert_equal(item1, item2, msg='Index %d%s' % (index, name))
```

As length of both lists are, same we need to check first list's element in 2nd list in case of ignore_order.

**Comment by bbpatel2001** [comment]
Hi @pekkaklarck, 
Can you please review following `_yield_list_diffs` function. 
need to import `from collections import Counter`
```
def lists_should_be_equal(list1, list2, ignore_order=False,msg=None, values=True,
                              names=None):
    ....
    diffs = list(_yield_list_diffs(list1, list2, names,ignore_order))
   ....

def _yield_list_diffs(list1, list2, names,ignore_order=False):
    **if ignore_order == True:
        assert_true(Counter(list1) == Counter(list2 ), msg='Lists  are  different')
    else:**
        for index, (item1, item2) in enumerate(zip(list1, list2)):         
            name = ' (%s)' % names[index] if index in names else ''
            try:
                assert_equal(item1, item2, msg='Index %d%s' % (index, name))
            except AssertionError as err:
                yield unic(err)
    ...
```
If we add these two lines before for loop in the existing given function will compare two UN-ordered  lists.

**Case 1:** 
['A', 'B', 'C', 'B']
['A', 'D', 'B', 'B', 'E']
lists_should_be_equal(my_list1,my_list2,ignore_order=True)
AssertionError: Lengths are different: 4 != 5 

**Case 2:** 
['A', 'B', 'C', 'B']
['A', 'D', 'B', 'B']
 lists_should_be_equal(my_list1,my_list2,ignore_order=True)
AssertionError: Lists  are  not matching

**Case 3:** 
['A', 'B', 'C', 'B']
['C', 'A', 'B', 'B']

PASS

Adding more tests and verifying it 
- Duplicates not removed and compares correctly. 
- No original list modified.

**Comment by bbpatel2001** [comment]
------------------------------------------------------------------------------
Lists Should Be Equal With Different Values Ignore_order :: FAIL L... | PASS |
------------------------------------------------------------------------------
Lists Should Be Equal With Different Values And Own Error Message ... | PASS |
------------------------------------------------------------------------------
Lists Should Be Equal With Different Values And Own And Default Er... | PASS |
------------------------------------------------------------------------------
Lists Should Be Equal With Named Indices As List Ignore_order :: F... | FAIL |
Lists are different:
Index 0 (a): 11 != 13
Index 2 (c): 13 != 11
------------------------------------------------------------------------------
Lists Should Be Equal With Named Indices As List With Too Few Valu... | PASS |
------------------------------------------------------------------------------
Lists Should Be Equal With Duplicate Value Ignore_order :: PASS li... | PASS |
------------------------------------------------------------------------------
Lists Should Be Equal With Duplicate Value Second Small List :: FA... | FAIL |
Lists  are  different
------------------------------------------------------------------------------
Lists Should Be Equal With Two Duplicate Value Ignore_order :: FAI... | PASS |
------------------------------------------------------------------------------
Lists Should Be Equal With Duplicate And Different Value Ignore_or... | FAIL |
Lists are different:
Index 2: A != E
Index 3: D != A

**Comment by bbpatel2001** [comment]
https://github.com/robotframework/robotframework/pull/2720

**Comment by bbpatel2001** [comment]
#2720

**Comment by pekkaklarck** [comment]
@bbpatel2001, I don't understand the usage of `collections.Counter` at all. I would assume just sorting lists before comparison would be the easiest solution. Are you interested to look at this again?

**Comment by chriscallan** [comment]
As Pekka notes, sorting the list contents and comparing those sorted lists is the easiest way to do this, and the way that I've handled this situation in other projects when I only care about list contents rather than order.

**Comment by kanchi240** [comment]
why not to use the "should be equal" keyword in buildin?

**Comment by sebastianciupinski** [comment]
What about introducing **Set** concept into framework itself and into Collections with related `Sets Should Be Equal` , `Remove From Set`, `Convert To Set`, etc?

**Comment by rainydew** [comment]
May this one help?
In our company I just extend this library to custom RF library and then import it to compare.

**Comment by bbpatel2001** [comment]
Hi @pekkaklarck could you please review PR https://github.com/robotframework/robotframework/pull/3469

**Comment by bbpatel2001** [comment]
is this PR #3469 RF 3.2 candidate?

**Comment by pekkaklarck** [comment]
This was implemented by PR #3469. Thanks a lot @bbpatel2001!
